mock/protocol.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mock
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/clearcontainers/proxy/api"
+)
+
+// XXX: could do with its own package to remove that ugly namespacing
+type protocolHandler func([]byte, interface{}, *handlerResponse)
+
+// Encapsulates the different parts of what a handler can return.
+type handlerResponse struct {
+	err     error
+	results map[string]interface{}
+}
+
+func (r *handlerResponse) SetError(err error) {
+	r.err = err
+}
+
+func (r *handlerResponse) SetErrorMsg(msg string) {
+	r.err = errors.New(msg)
+}
+
+func (r *handlerResponse) SetErrorf(format string, a ...interface{}) {
+	r.SetError(fmt.Errorf(format, a...))
+}
+
+func (r *handlerResponse) AddResult(key string, value interface{}) {
+	if r.results == nil {
+		r.results = make(map[string]interface{})
+	}
+	r.results[key] = value
+}
+
+type FrameKey struct {
+	ftype  api.FrameType
+	opcode int
+}
+
+func newFrameKey(frameType api.FrameType, opcode int) FrameKey {
+	return FrameKey{
+		ftype:  frameType,
+		opcode: opcode,
+	}
+}
+
+type protocol struct {
+	cmdHandlers map[FrameKey]protocolHandler
+}
+
+func newProtocol() *protocol {
+	return &protocol{
+		cmdHandlers: make(map[FrameKey]protocolHandler),
+	}
+}
+
+func (proto *protocol) Handle(key FrameKey, handler protocolHandler) bool {
+	if _, ok := proto.cmdHandlers[key]; ok {
+		return false
+	}
+	proto.cmdHandlers[key] = handler
+	return true
+}
+
+type clientCtx struct {
+	conn net.Conn
+
+	userData interface{}
+}
+
+func newErrorResponse(opcode int, errMsg string) *api.Frame {
+	frame, err := api.NewFrameJSON(api.TypeResponse, opcode, &api.ErrorResponse{
+		Message: errMsg,
+	})
+	if err != nil {
+		frame, err = api.NewFrameJSON(api.TypeResponse, opcode, &api.ErrorResponse{
+			Message: fmt.Sprintf("couldn't marshal response: %v", err),
+		})
+	}
+	if err != nil {
```

Comment:
You should not test the err here, this is confusing. Embed this test inside the previous statement, or you can use reversed logic (test err == nil) in order to avoid too much indentation in your code.